Azure Redis Cache AAD OSS Client Library Samples (#28732)
Azure/azure-sdk-for-java: sdk/identity/azure-identity/src/samples/Azure Cache For Redis/Lettuce/Azure-AAD-Authentication-With-Lettuce.md

## Azure Cache for Redis: Azure AD with Lettuce client library

### Table of contents

- [Dependency Requirements](#dependency-requirements)
- [Authenticate with Azure AD - Hello World](#authenticate-with-azure-ad-hello-world)

> - [Authenticate with Azure AD - Handle Re-Authentication](#authenticate-with-azure-ad-handle-re-authentication)

Reviewer: ```suggestion
- [Authenticate with Azure AD - Handle Reauthentication](#authenticate-with-azure-ad-handle-reauthentication)
```

#### Dependency Requirements
```xml
<dependency>
    <groupId>com.azure</groupId>
    <artifactId>azure-identity</artifactId>
    <version>1.5.0</version>
</dependency>

<dependency>
    <groupId>io.lettuce</groupId>
    <artifactId>lettuce-core</artifactId>
    <version>6.1.8.RELEASE</version>
</dependency>
```

#### Samples Guidance
Familiarity with the [Lettuce](https://github.com/lettuce-io/lettuce-core) and [Azure Identity for Java](https://docs.microsoft.com/azure/developer/java/sdk/identity) client libraries is assumed.

[Authenticate with Azure AD - Hello World](#authenticate-with-azure-ad-hello-world)
This sample is recommended for users getting started to use Azure AD authentication with Azure Cache for Redis.

> [Authenticate with Azure AD - Handle Re-Authentication](#authenticate-with-azure-ad-handle-re-authentication)

Reviewer: ```suggestion
[Authenticate with Azure AD - Handle Reauthentication](#authenticate-with-azure-ad-handle-reauthentication)
```

> This sample is recommended users looking to build long-running applications and would like to handle re authenticating with Azure AD upon token expiry.

Reviewer: ```suggestion
This sample is recommended to users looking to build long-running applications and would like to handle reauthenticating with Azure AD upon token expiry.
```

**Note:** The below sample uses the Azure Identity library's `ClientCertificateCredential`. The credential can be replaced with any of the other `TokenCredential` implementations offered by the Azure Identity library.

##### Challenges

> The Lettuce Redis Client Handshake process is described below. The RESP 3 Protocol uses `HELLO` command which currently is not supported on service side.

Reviewer: ```suggestion
The Lettuce client handshake process is described below. The RESP 3 Protocol uses `HELLO` command, which currently isn't supported on the service side.
```

```java

// Lettuce RESP 2 Protocol.

walmart-demo.redis.cache.windows.net:6379> AUTH walmartdummyuser eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6ImpTMVhvMU9XRGpfNTJ2YndHTmd2UU8yVnpNYyIsImtpZCI6ImpTMVhvMU9XRGpfNTJ2YndHTmd2UU8yVnpNYyJ9.eyJhdWQiOiJodHRwczovLyouY2FjaGVpbmZyYS53aW5kb3dzLm5ldDoxMDIyNS9hcHBpZCIsImlzcyI6Imh0dHBzOi8vc3RzLndpbmRvd3MubmV0LzcyZjk4OGJmLTg2ZjEtNDFhZi05MWFiLTJkN2NkMDExZGI0Ny8iLCJpYXQiOjE2NTA4OTkxMzUsIm5iZiI6MTY1MDg5OTEzNSwiZXhwIjoxNjUwOTg1ODM1LCJhaW8iOiJFMlpnWU1pWmVIWVpZOC8xRHhjdjl2ZlhPKzdjQWdBPSIsImFwcGlkIjoiZmM1.....

OK

// Lettuce RESP 3 Protocol.
walmart-demo.redis.cache.windows.net:6379> HELLO 3 AUTH walmartdummyuser eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6ImpTMVhvMU9XRGpfNTJ2YndHTmd2UU8yVnpNYyIsImtpZCI6ImpTMVhvMU9XRGpfNTJ2YndHTmd2UU8yVnpNYyJ9.eyJhdWQiOiJodHRwczovLyouY2FjaGVpbmZyYS53aW5kb3dzLm5ldDoxMDIyNS9hcHBpZCIsImlzcyI6Imh0dHBzOi8vc3RzLndpbmRvd3MubmV0LzcyZjk4OGJmLTg2ZjEtNDFhZi05MWFiLTJkN2NkMDExZGI0Ny8iLCJpYXQiOjE2NTA4OTkxMzUsIm5iZiI6MTY1MDg5OTEzNSwiZXhwIjoxNjUwOTg1ODM1LCJhaW8iOiJFMlpnWU1pWmVIWVpZOC8xRHhjdjl2ZlhPKzdjQWdBPSIsImFwcGlkIjoi.....

(error) WRONGPASS invalid username-password pair

```

##### Authenticate with Azure AD: Hello World

> This sample is intended to assist in authenticating with Azure AD via the Lettuce client library. It focuses on displaying the logic required to fetch an Azure AD access token and to use it as password when setting up the Lettuce Redis client instance.

Reviewer: ```suggestion
This sample is intended to assist in authenticating with Azure AD via the Lettuce client library. It focuses on displaying the logic required to fetch an Azure AD access token and to use it as password when setting up the Lettuce client instance.
```

##### Raw Redis CLI Connection Behavior
```java

// Lettuce RESP 2 Protocol.

walmart-demo.redis.cache.windows.net:6379> AUTH walmartdummyuser eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6ImpTMVhvMU9XRGpfNTJ2YndHTmd2UU8yVnpNYyIsImtpZCI6ImpTMVhvMU9XRGpfNTJ2YndHTmd2UU8yVnpNYyJ9.eyJhdWQiOiJodHRwczovLyouY2FjaGVpbmZyYS53aW5kb3dzLm5ldDoxMDIyNS9hcHBpZCIsImlzcyI6Imh0dHBzOi8vc3RzLndpbmRvd3MubmV0LzcyZjk4OGJmLTg2ZjEtNDFhZi05MWFiLTJkN2NkMDExZGI0Ny8iLCJpYXQiOjE2NTA4OTkxMzUsIm5iZiI6MTY1MDg5OTEzNSwiZXhwIjoxNjUwOTg1ODM1LCJhaW8iOiJFMlpnWU1pWmVIWVpZOC8xRHhjdjl2ZlhPKzdjQWdBPSIsImFwcGlkIjoiZmM1.....

OK

// Lettuce RESP 3 Protocol.
walmart-demo.redis.cache.windows.net:6379> HELLO 3 AUTH walmartdummyuser eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6ImpTMVhvMU9XRGpfNTJ2YndHTmd2UU8yVnpNYyIsImtpZCI6ImpTMVhvMU9XRGpfNTJ2YndHTmd2UU8yVnpNYyJ9.eyJhdWQiOiJodHRwczovLyouY2FjaGVpbmZyYS53aW5kb3dzLm5ldDoxMDIyNS9hcHBpZCIsImlzcyI6Imh0dHBzOi8vc3RzLndpbmRvd3MubmV0LzcyZjk4OGJmLTg2ZjEtNDFhZi05MWFiLTJkN2NkMDExZGI0Ny8iLCJpYXQiOjE2NTA4OTkxMzUsIm5iZiI6MTY1MDg5OTEzNSwiZXhwIjoxNjUwOTg1ODM1LCJhaW8iOiJFMlpnWU1pWmVIWVpZOC8xRHhjdjl2ZlhPKzdjQWdBPSIsImFwcGlkIjoi.....

(error) WRONGPASS invalid username-password pair

```

> ##### Authenticate with Azure AD Hello World

Reviewer: ```suggestion
##### Authenticate with Azure AD: Hello World
```

        private TokenCredential tokenCredential;
        private final String username;

        /**
         * Create instance of Azure Redis Credentials
         * @param username the username to be used for authentication.
         * @param tokenCredential the token credential to be used to fetch requests.
         */
        public AzureRedisCredentials(String username, TokenCredential tokenCredential) {
            Objects.requireNonNull(username, "Username is required");
            Objects.requireNonNull(tokenCredential, "Token Credential is required");
            this.username = username;
            this.tokenCredential = tokenCredential;
        }

        @Override
        public String getUsername() {
            return username;
        }

        @Override
        public boolean hasUsername() {
            return username != null;
        }

        @Override
        public char[] getPassword() {
            return tokenCredential
                    .getToken(tokenRequestContext).block().getToken()
                    .toCharArray();
        }

        @Override
        public boolean hasPassword() {
            return tokenCredential != null;
        }
    }
```

> ##### Authenticate with Azure AD Handle Re Authentication.

Reviewer: ```suggestion
##### Authenticate with Azure AD: Handle Reauthentication
```

## Azure Cache For Redis AAD With Lettuce Client Library

### Table of contents

- [Lettuce Library](#lettuce-library)
    - [Dependency Requirements](#dependency-requirements)
    - [Authenticate with AAD - Hello World](#authenticate-with-aad-hello-world)
    - [Authenticate with AAD - Handle Re-Authentication](#authenticate-with-aad-handle-re-authentication)
    - [Authenticate with AAD - Azure Lettuce Client Wrapper](#authenticate-with-aad-azure-lettuce-wrapper)

### Lettuce Library

#### Dependency Requirements
```
        <dependency>
            <groupId>com.azure</groupId>
            <artifactId>azure-identity</artifactId>
            <version>1.5.0</version>
        </dependency>
        
        <dependency>
            <groupId>io.lettuce</groupId>
            <artifactId>lettuce-core</artifactId>
            <version>6.1.8.RELEASE</version>
        </dependency>

```

##### Challenges

> ![image](https://user-images.githubusercontent.com/5430778/165149837-370c340a-7bf4-4067-a8a7-3a1a7e049773.png)

Reviewer: Store the image in this repo.

Familiarity with the Jedis and Azure Identity client libraries is assumed. If you're new to the Azure Identity library for Java, see the docs for [Azure Identity](https://docs.microsoft.com/azure/developer/java/sdk/identity) and [Lettuce](https://www.javadoc.io/doc/redis.clients/jedis/latest/index.html) rather than this guide.

##### Migration Guidance
When migrating your existing your application code, you need to replace the password input with Azure Active Directory Token.
Integrate the logic in your application code to fetch an AAD Access Token via Identity SDK as shown below and replace the password configuring/retrieving logic in your application code.

**Note:** The below sample uses `ClientCertificateCredential` from our [Azure Identity](https://docs.microsoft.com/azure/developer/java/sdk/identity) SDK, the credential can be replaced with any of the other `TokenCredential` implementations offered by our [Azure Identity](https://docs.microsoft.com/azure/developer/java/sdk/identity) SDK.

##### Version 6.1.8.RELEASE or less

> ```

Reviewer: ```suggestion
```xml
```

                .withPort(6379)
                .withSsl(false) // Targeting Non-SSL 6379 port.
                .withAuthentication("USERNAME", token)
                .withClientName("LettuceClient")
                .build();

        // Create Lettuce Redis Client
        RedisClient client = RedisClient.create(redisURI);

        // Configure the client options.
        client.setOptions(ClientOptions.builder()
                .socketOptions(SocketOptions.builder()
                        .keepAlive(true) // Keep the connection alive to work with Azure Redis Cache
                        .build())
                .protocolVersion(ProtocolVersion.RESP2) // Use RESP2 Protocol to ensure AUTH command is used for handshake.
                .build());

        StatefulRedisConnection<String, String> connection = client.connect(StringCodec.UTF8);

        // Create the connection, in this case we're using a sync connection, but you can create async / reactive connections as needed.
        RedisStringCommands sync = connection.sync();
        sync.set("Az:testKey", "testVal");
        System.out.println(sync.get("Az:testKey").toString());
    }

```

##### Version 6.2.0-BUILD-SNAPSHOT

> ```

Reviewer: ```suggestion
```xml
```

Familiarity with the Jedis and Azure Identity client libraries is assumed. If you're new to the Azure Identity library for Java, see the docs for [Azure Identity](https://docs.microsoft.com/azure/developer/java/sdk/identity) and [Jedis](https://www.javadoc.io/doc/redis.clients/jedis/latest/index.html) rather than this guide.

##### Migration Guidance
When migrating your existing your application code, you need to replace the password input with Azure Active Directory Token.
Integrate the logic in your application code to fetch an AAD Access Token via Identity SDK as shown below and replace the password configuring/retrieving logic in your application code.

**Note:** The below sample uses `ClientCertificateCredential` from our [Azure Identity](https://docs.microsoft.com/azure/developer/java/sdk/identity) SDK, the credential can be replaced with any of the other `TokenCredential` implementations offered by our [Azure Identity](https://docs.microsoft.com/azure/developer/java/sdk/identity) SDK.

##### Version 6.1.8.RELEASE or less

> ```

Reviewer: ```suggestion
```xml
```

        // Build Redis URI with host and authentication details.
        RedisURI redisURI = RedisURI.Builder.redis(hostName)
                .withPort(port)
                .withSsl(false) // Targeting Non-SSL 6379 port.
                .withAuthentication(username, accessToken.getToken())
                .withClientName("LettuceClient")
                .build();

        // Create Lettuce Redis Client
        RedisClient client = RedisClient.create(redisURI);

        // Configure the client options.
        client.setOptions(ClientOptions.builder()
                .socketOptions(SocketOptions.builder()
                        .keepAlive(true) // Keep the connection alive to work with Azure Redis Cache
                        .build())
                .protocolVersion(ProtocolVersion.RESP2) // Use RESP2 Protocol to ensure AUTH command is used for handshake.
                .build());

        return client;
    }

    private static AccessToken getAccessToken(TokenCredential tokenCredential, TokenRequestContext trc) {
        return tokenCredential.getToken(trc).block();
    }
```

##### Version 6.2.0-BUILD-SNAPSHOT

> ```

Reviewer: ```suggestion
```xml
```

                .build())
                .protocolVersion(ProtocolVersion.RESP2) // Use RESP2 Protocol to ensure AUTH command is used for handshake.
        .build());

StatefulRedisConnection<String, String> connection = client.connect(StringCodec.UTF8);

// Create the connection, in this case we're using a sync connection, but you can create async / reactive connections as needed.
RedisStringCommands sync = connection.sync();
sync.set("Az:testKey", "testVal");
System.out.println(sync.get("Az:testKey").toString());

```

With `AzureLetttuceRedisClient`, client instances are created via builders. The builder accepts the:

- RedisURI
- Token Credential object that's used to generate a token
- Client Options
- Client Resources

See the following example of setting up the Azure Jedis client.

**Note:** The below sample uses `ClientCertificateCredential` from our [Azure Identity](https://docs.microsoft.com/azure/developer/java/sdk/identity) SDK, the credential can be replaced with any of the other `TokenCredential` implementations offered by our [Azure Identity](https://docs.microsoft.com/azure/developer/java/sdk/identity) SDK.

> ![image](https://user-images.githubusercontent.com/5430778/166561705-027b3fa7-384b-4939-a95f-ca75b6134d4f.png)

Reviewer: Store this image in the repo

##### Version 6.1.8.RELEASE or less

> ```

Reviewer: ```suggestion
```xml
```

                        .build())
                .tokenCredential(clientCertificateCredential)
                .username("USERNAME")
                .clientOptions(ClientOptions.builder()
                        .socketOptions(SocketOptions.builder()
                                .keepAlive(true)
                                .build())
                        .protocolVersion(ProtocolVersion.RESP2)
                        .build())
                .build();

        StatefulRedisConnection<String, String> connection = client.connect(StringCodec.UTF8);

        RedisStringCommands sync = connection.sync();
        try {
            sync.set("Az:testKey", "testVal");
            System.out.println(sync.get("Az:testKey").toString());
        } catch (RedisException e) {
            // Handle Exception as required.

            // Check if connection open, else open a new connection instance for further use.
            if (!connection.isOpen()) {
                connection = client.connect(StringCodec.UTF8);
                sync = connection.sync();
            }
        }
    }
```

##### Lettuce Version 6.2.0-BUILD-SNAPSHOT

> ```

Reviewer: ```suggestion
```xml
```
